src/cryptonote_core/service_node_list.cpp

```diff
@@ -1457,6 +1457,14 @@ namespace service_nodes
       }
       else if (type == quorum_type::checkpointing)
       {
+        // Checkpoint quorums only exist every CHECKPOINT_INTERVAL blocks, but the height that gets
+        // used to generate the quorum (i.e. the `height` variable here) is actually `H -
+        // REORG_SAFETY_BUFFER_BLOCKS_POST_HF12`, where H is divisible by CHECKPOINT_INTERVAL, but
+        // REORG_SAFETY_BUFFER_BLOCKS_POST_HF12 is not (it equals 11).  Hence the addition here to
+        // "undo" the lag before checking to see if we're on an interval multiple:
+        if ((height + REORG_SAFETY_BUFFER_BLOCKS_POST_HF12) % CHECKPOINT_INTERVAL != 0)
```

Comment:
It shouldn't *ever* trigger the "was not cached" error: basically asking for a checkpointing quorum on an invalid height is a bug in the first place, and I think if something does call it we *want* that error (it already helped me track down a related issue, fixed in this PR).

Reply:
I still don't think we want to make it so that supplying a non (height + REORG_SAFETY_BUFFER_BLOCKS_POST_HF12) % CHECKPOINT_INTERVAL == 0) height to get_testing_quorum to be a nullptr if you are querying the checkpointing quorum. It should just return an empty quorum because we have no work to be done on that height. It adds an error for a very specific combination of inputs. 

The log can be moved into get_testing_quorum that querying the height is unexpected but still allowed and then all the same error checking behaviour can be maintained.

The height iteration with m_last_checkpointed_height needs to be updated as it still goes += CHECKPOINT_INTERVAL (4) but then I subtract REORG_SAFETY_BUFFER_BLOCKS (11) which is always going to make you end up on a non multiple of 4.

Reply:
> The height iteration with m_last_checkpointed_height needs to be updated as it still goes += CHECKPOINT_INTERVAL (4) but then I subtract REORG_SAFETY_BUFFER_BLOCKS (11) which is always going to make you end up on a non multiple of 4.

This is actually correct: checkpointing quorums are at N+1 where N is divisible by 4.

Reply:
(That is a little weird, but changing it would break any existing 4.0.2 clients, while the N+1 approach keeps it compatible).

Reply:
> > The height iteration with m_last_checkpointed_height needs to be updated as it still goes += CHECKPOINT_INTERVAL (4) but then I subtract REORG_SAFETY_BUFFER_BLOCKS (11) which is always going to make you end up on a non multiple of 4.
> 
> This is actually correct: checkpointing quorums are at N+1 where N is divisible by 4.

oh right, got it.

Reply:
> I still don't think we want to make it so that supplying a non (height + REORG_SAFETY_BUFFER_BLOCKS_POST_HF12) % CHECKPOINT_INTERVAL == 0) height to get_testing_quorum to be a nullptr if you are querying the checkpointing quorum. It should just return an empty quorum because we have no work to be done on that height. It adds an error for a very specific combination of inputs.

I did consider both approaches, and I still believe that a null pointer signals an invalid quorum better than an empty vector--and basically all the code that is calling this is doing the right thing.  For example, it requires no change to the RPC code which is doing exactly the right thing (skipping nullptr quorums), and is properly handled by the checkpoint code which (correctly) prints an error if an expected quorum doesn't exist.  But it *can't* raise such an error for an empty quorum because an empty quorum *is* a valid quorum in a way that a null quorum isn't.

Reply:
Yeah, in my view I don't really see it as an error to get an empty array so I prefer that way. With returning null for it, it makes using it more brittle when getting checkpointing quorums. But there's no special behaviour needed either way since you'll always be checking for a nullptr and you would always be iterating/copying an empty array which pretty much always works in both cases.

In any case with this, the height needing to be modulo N needs to be documented at the function prototype.

Reply:
> But there's no special behaviour needed either way since you'll always be checking for a nullptr and you would always be iterating/copying an empty array which pretty much always works in both cases.

Except it isn't true: if you change it you will also have to change the RPC to skip returning empty quorums because without this, there is a bug where it returns a checkpoint quorum for every single height in the request range, which just seems wrong (and pushes dealing with the no-quorum state onto the RPC consumer).  That means that the block explorer now has to guess whether an empty quorum is a *missing* quorum or actually an empty quorum (because, for instance, we don't have any nodes or enough nodes to make up a quorum).

> Yeah, in my view I don't really see it as an error to get an empty array so I prefer that way.

I quite strongly disagree.  I think you're overwriting an error condition (the quorum is not defined) with a non-error condition (the quorum is defined and is empty).  However, if you feel strongly enough about it, feel free to implement it over my objections.

Reply:
I'm ok to agree to disagree, we could go on forever.

Reply:
> In any case with this, the height needing to be modulo N needs to be documented at the function prototype.

With you 100% on this one though.

Reply:
Just added in another PR, gonna merge this one